File: Top200/199-RightSideView.java
package Top200;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class RightSideView {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }

    public List<Integer> rightSideView(TreeNode root) {        
        List<Integer> res = new ArrayList<Integer>();
        if (root == null) {
            return res;
        }

        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            TreeNode temp = null;

            for (int i = 0; i < size; i++) {
                temp = queue.poll();
                
                if (temp.left != null) {
                    queue.offer(temp.left);
                }

                if (temp.right != null) {
                    queue.offer(temp.right);
                }
                
                if (i == size - 1) {
                    res.add(temp.val);
                }
            }
        }

        return res;
    }

    public int maxDepth = 0;
    public int currDepth = 0;
    public List<Integer> rightSideView2(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();

        dfs(root, res);
        
        return res;
    }

    public void dfs(TreeNode root, List<Integer> res) {
        if (root == null) {
            return;
        }

        ++ currDepth;
        if (currDepth > maxDepth) {
            maxDepth = currDepth;
            res.add(root.val);
        }

        dfs(root.right, res);
        dfs(root.left, res);

        currDepth --;
    }

    public List<Integer> rightSideView3(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();

        dfs(root, res, 0);

        return res;
    }

    public void dfs(TreeNode root, List<Integer> res, int depth) {
        if (root == null) {
            return;
        }

        if (depth == res.size()) {
            res.add(root.val);
        }

        ++ depth;

        dfs(root.right, res, depth);
        dfs(root.left, res, depth);
    }
    
}

File: Top200/199-RightSideView.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Queue struct {
	arr  []*TreeNode
	size int
}

func NewQueue() *Queue {
	return &Queue{
		arr:  []*TreeNode{},
		size: 0,
	}
}

func (q *Queue) Pop() *TreeNode {
	if q.size <= 0 {
		return nil
	}

	res := q.arr[0]
	q.arr = q.arr[1:q.size]
	q.size--

	return res
}

func (q *Queue) Push(node *TreeNode) {
	q.arr = append(q.arr, node)
	q.size++
}

func (q *Queue) IsEmpty() bool {
	if q.size <= 0 {
		return true
	}

	return false
}

func (q *Queue) Size() int {
	return q.size
}

func rightSideView(root *TreeNode) []int {
	res := []int{}
	queue := NewQueue()

	if root == nil {
		return res
	}

	queue.Push(root)

	for !queue.IsEmpty() {
		size := queue.Size()
		res = append(res, queue.arr[size-1].Val)

		for i := 0; i < size; i++ {
			node := queue.Pop()

			if node.Left != nil {
				queue.Push(node.Left)
			}

			if node.Right != nil {
				queue.Push(node.Right)
			}
		}
	}

	return res
}
